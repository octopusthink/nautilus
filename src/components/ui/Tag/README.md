`<Strong>` is used to mark a short span of text within body copy as important.

```jsx
import { Heading, Paragraph } from '@octopusthink/nautilus';

<React.Fragment>
  <Heading>Wonder Women</Heading>
  <Paragraph>Once upon a time there were a whole group of lady-types who lived in a beautiful paradise-land called ... well, I forget but it was really pretty and just chock full of badass ladies.</Paragraph>
  <Tag color="hotpink">Amazon</Tag><Tag color="hotpink">Superhero</Tag><Tag color="hotpink">Ladies who lunch</Tag>

</React.Fragment>

```


```jsx
<React.Fragment>
  <Tag color="hotpink">Hot pink</Tag>
  <Tag color="navy">Navy</Tag>
  <Tag color="aqua">Aqua</Tag>
  <Tag color="#053e6c">Blue</Tag>
  <Tag color="#8a0000">Red</Tag>
  <Tag color="#eeb600">Yellow</Tag>
</React.Fragment>
```


```jsx
import { Heading, Paragraph } from '@octopusthink/nautilus';

<React.Fragment>
  <Tag>12 September 2019</Tag>
  <Heading>Wonder Women</Heading>
</React.Fragment>

```
