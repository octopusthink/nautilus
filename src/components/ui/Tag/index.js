import { css } from '@emotion/core';
import styled from '@emotion/styled';
import PropTypes from 'prop-types';
import React, { forwardRef } from 'react';

import { Icon } from 'components/ui/Icon';
import { getContrastingTextColor, metadata, toUnits } from 'styles';
import { useTheme } from 'themes';

export const Tag = forwardRef((props, ref) => {
  const theme = useTheme();
  const {
    badge,
    children,
    color,
    label,
    onDismiss,
    status,
    ...otherProps
  } = props;

  return (
    <span ref={ref} {...otherProps}>
      {label && (
        <span // TODO: switch with VisuallyHidden component, once available.
          css={css`
            font-style: bold;
            position: absolute;
            left: -10000px;
            top: auto;
            width: 1px;
            height: 1px;
            overflow: hidden;
          `}
        >
          {label}:
        </span>
      )}

      {children}
      {onDismiss && (
        <button
          type="button"
          onClick={onDismiss}
          css={css`
            box-shadow: 0;
            border: 0;
            background: none;
            color: currentColor;
            cursor: pointer;
            vertical-align: 0.2rem;
            margin: 0;
            padding: 0;

            &:hover {
              opacity: 0.75;
            }
          `}
        >
          <Icon
            name="x"
            small
            css={css`
              display: inline-block;
              margin-left: ${toUnits(theme.spacing.padding.small)};
              margin-top: ${toUnits(theme.spacing.padding.xxSmall)};
              opacity: 1;
              vertical-align: middle;
            `}
          />
        </button>
      )}
    </span>
  );
});

export const styles = (props) => {
  const { color, badge, status, theme } = props;
  let textColor = theme.colors.text.light;
  let backgroundColor;

  if (color) {
    backgroundColor = color;
  }

  if (status) {
    backgroundColor = theme.colors.intent[status];
  }

  if (status || color) {
    textColor = getContrastingTextColor({ color: backgroundColor, theme });
  }

  return css`
    ${metadata.small(theme)};
    color: ${textColor};
    margin: 0 ${toUnits(theme.spacing.padding.xSmall)}
      ${toUnits(theme.spacing.padding.xSmall)} 0;

    ${backgroundColor &&
      css`
        background: ${backgroundColor};
        padding: ${toUnits(theme.spacing.padding.xSmall)}
          ${toUnits(theme.spacing.padding.small)};
      `}

    ${badge &&
      css`
        border-radius: ${toUnits(theme.spacing.padding.xLarge)};
        min-width: ${toUnits(theme.spacing.padding.xLarge)};
        height: ${toUnits(theme.spacing.padding.xLarge)};
        display: inline-flex;
        align-items: center;
        justify-content: center;
        box-sizing: border-box;
      `}
  `;
};

Tag.defaultProps = {
  children: undefined,
  color: null,
  label: null,
  badge: false,
  onDismiss: null,
  status: undefined,
};

Tag.propTypes = {
  /** @ignore */
  children: PropTypes.node,
  /** Use colour to differentiate different tags or indicate status. */
  color: PropTypes.string,
  /** Optional label for providing additional context to screen reader users. */
  label: PropTypes.string,
  /** Style numbers or counts with a badge. */
  badge: PropTypes.bool,
  /** Function to call when a Tag is dismissed via the close button. */
  onDismiss: PropTypes.func,
  /** Indicate status using a semantic colour set. */
  status: PropTypes.oneOf([
    'neutral',
    'success',
    'warning',
    'danger',
    'inProgress',
    'new',
  ]),
};

export const { defaultProps, propTypes } = Tag;

Tag.displayName = 'Tag';

export default styled(Tag)(styles);
