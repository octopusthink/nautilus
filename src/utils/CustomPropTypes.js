import { Children } from 'react';

const allowedChildren = (...types) => {
  return (propValue, key, componentName) => {
    let error = null;

    Children.forEach(propValue[key], (child) => {
      if (!types.includes(child.type)) {
        error = new Error(
          `${componentName} component only accepts the following components as children: ${types
            .map((element) => element.displayName || element)
            .join(', ')}.`,
        );
      }
    });

    return error;
  };
};

export default {
  allowedChildren,
};
