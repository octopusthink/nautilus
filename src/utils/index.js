export { default as CustomPropTypes } from './CustomPropTypes';
