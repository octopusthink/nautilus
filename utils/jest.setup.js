// import axios from 'axios';
// import AxiosMockAdapter from 'axios-mock-adapter';
// import { JSDOM } from 'jsdom';
// eslint-disable-next-line import/no-extraneous-dependencies
import 'jest-axe/extend-expect';
// eslint-disable-next-line import/no-extraneous-dependencies
import '@testing-library/jest-dom/extend-expect';

global.window.matchMedia = jest.fn().mockImplementation((query) => ({
  matches: false,
  media: query,
  onchange: null,
  addListener: jest.fn(),
  removeListener: jest.fn(),
}));

jest.spyOn(global.console, 'error');
jest.spyOn(global.console, 'warn');
jest.spyOn(global.console, 'log');
jest.spyOn(global.console, 'info');
jest.spyOn(global.console, 'debug');

beforeEach(() => {
  global.console.error.mockClear();
  global.console.warn.mockClear();
  global.console.log.mockClear();
  global.console.info.mockClear();
  global.console.debug.mockClear();
});
