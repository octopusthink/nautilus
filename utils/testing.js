// We only use react-testing-library in development, so ignore this linting
// error.
// eslint-disable-next-line import/no-extraneous-dependencies
import { render } from '@testing-library/react';
// eslint-disable-next-line import/no-extraneous-dependencies
import { createMemoryHistory } from 'history';
import PropTypes from 'prop-types';
import React from 'react';
import { Router } from 'react-router-dom';

import { Nautilus } from 'components';
import { nautilus as nautilusDefaultTheme, themePropTypes } from 'themes';

// Wrap `react-testing-library`'s `render` function with our providers.
const NautilusProviders = ({ children, theme }) => {
  return <Nautilus theme={theme}>{children}</Nautilus>;
};

NautilusProviders.defaultProps = {
  children: undefined,
  theme: nautilusDefaultTheme,
};

NautilusProviders.propTypes = {
  /** @ignore */
  children: PropTypes.node,
  /** @ignore */
  theme: themePropTypes,
};

const customRender = (ui, options) => {
  return render(ui, { wrapper: NautilusProviders, ...options });
};

// this is a handy function that I would utilize for any component
// that relies on the router being in context
export const renderWithRouter = (
  ui,
  {
    route = '/',
    history = createMemoryHistory({ initialEntries: [route] }),
  } = {},
) => {
  return {
    ...customRender(<Router history={history}>{ui}</Router>),
    // adding `history` to the returned utilities to allow us
    // to reference it in our tests (just try to avoid using
    // this to test implementation details).
    history,
  };
};

export const muteConsole = ({ times, type } = {}) => {
  Array.from({ length: times }).forEach(() => {
    global.console[type].mockImplementationOnce(() => jest.fn());
  });
};

// Export jest-axe helpers.
export { axe } from 'jest-axe';

// Re-export everything else as normal.
export * from '@testing-library/react';

// override render method
export { customRender as render };
