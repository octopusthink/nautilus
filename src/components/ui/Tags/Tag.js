import { css } from '@emotion/core';
import PropTypes from 'prop-types';
import React, { useState } from 'react';

import { Icon } from 'components/ui/Icon';
import { getContrastingTextColor, metadata, toUnits } from 'styles';
import { useTheme } from 'themes';

export const Tag = (props) => {
  const [isDismissed, setDismissed] = useState(false);
  const theme = useTheme();
  const { badge, children, color, onDismiss, status, ...otherProps } = props;

  if (isDismissed) {
    return null;
  }

  let textColor = theme.colors.text.light;
  let backgroundColor;
  if (color) {
    backgroundColor = color;
  }

  if (status) {
    backgroundColor = theme.colors.intent[status];
  }

  if (status || color) {
    textColor = getContrastingTextColor({ color: backgroundColor, theme });
  }

  return (
    <span
      css={css`
        ${metadata.small(theme)};
        color: ${textColor};
        margin: 0 ${toUnits(theme.spacing.padding.xSmall)} ${toUnits(theme.spacing.padding.xSmall)}
          0;

        ${backgroundColor &&
          css`
            background: ${backgroundColor};
            padding: ${toUnits(theme.spacing.padding.xSmall)}
              ${toUnits(theme.spacing.padding.small)};
          `}

        ${badge &&
          css`
            border-radius: ${toUnits(theme.spacing.padding.xLarge)};
            min-width: ${toUnits(theme.spacing.padding.xLarge)};
            height: ${toUnits(theme.spacing.padding.xLarge)};
            display: inline-flex;
            align-items: center;
            justify-content: center;
            box-sizing: border-box;
          `}
      `}
      {...otherProps}
    >
      {children}
      {onDismiss && (
        <button
          css={css`
            box-shadow: 0;
            border: 0;
            background: none;
            color: currentColor;
            cursor: pointer;
            vertical-align: 0.2rem;
            margin: 0;
            padding: 0;

            &:hover {
              opacity: 0.75;
            }
          `}
          onClick={(event) => {
            event.preventDefault();

            setDismissed(true);

            onDismiss(event);
          }}
          type="button"
        >
          <Icon
            name="x"
            small
            css={css`
              display: inline-block;
              margin-left: ${toUnits(theme.spacing.padding.small)};
              margin-top: ${toUnits(theme.spacing.padding.xxSmall)};
              opacity: 1;
              vertical-align: middle;
            `}
          />
        </button>
      )}
    </span>
  );
};

Tag.defaultProps = {
  children: undefined,
  color: null,
  badge: false,
  onDismiss: null,
  status: undefined,
};

Tag.propTypes = {
  /** @ignore */
  children: PropTypes.node,
  /** Use colour to differentiate different tags or indicate status. */
  color: PropTypes.string,
  /** Style numbers or counts with a badge. */
  badge: PropTypes.bool,
  /** Function to call when a Tag is dismissed via the close button. */
  onDismiss: PropTypes.func,
  /** Indicate status using a semantic colour set. */
  status: PropTypes.oneOf(['neutral', 'success', 'warning', 'danger', 'inProgress', 'new']),
};

export const { defaultProps, propTypes } = Tag;

export default Tag;
