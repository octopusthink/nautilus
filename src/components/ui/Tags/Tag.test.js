import React from 'react';

import { fireEvent, render } from 'utils/testing';

import Tag from './Tag';

describe('Tag', () => {
  it('should render a <span> tag', () => {
    const { container } = render(<Tag>Hello</Tag>);

    expect(container.firstChild.tagName).toEqual('SPAN');
  });

  it('should output its children', () => {
    const { getByTestId } = render(
      <Tag>
        <span data-testid="child" />
      </Tag>,
    );

    expect(getByTestId('child')).toBeDefined();
  });

  it('should accept and pass through other props', () => {
    const { getByTestId } = render(
      <Tag className="custom-class" data-testid="myText">
        hello
      </Tag>,
    );

    expect(getByTestId('myText').classList).toContain('custom-class');
  });

  it('should match styles', () => {
    const { container } = render(<Tag>hello</Tag>);

    expect(container).toMatchSnapshot();
  });

  it('should dismiss a Tag with `onDismiss` set', () => {
    const onDismiss = jest.fn();
    const { container, getByTestId } = render(
      <Tag data-testid="myTag" onDismiss={onDismiss}>
        hello
      </Tag>,
    );

    expect(getByTestId('myTag')).toBeDefined();

    fireEvent.click(container.querySelector('button'));

    expect(onDismiss).toHaveBeenCalled();
    expect(() => {
      getByTestId('myTag');
    }).toThrow('Unable to find an element');
  });
});
