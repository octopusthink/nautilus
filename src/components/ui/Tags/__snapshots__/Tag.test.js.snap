// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tag should match styles 1`] = `
.emotion-0 {
  font-family: -apple-system,BlinkMacSystemFont,San Francisco,Roboto,Segoe UI,Helvetica Neue,sans-serif;
  font-weight: 700;
  font-size: 1.5rem;
  line-height: 1.3333333333333333;
  -webkit-letter-spacing: 0.05em;
  -moz-letter-spacing: 0.05em;
  -ms-letter-spacing: 0.05em;
  letter-spacing: 0.05em;
  text-transform: uppercase;
  color: #666c76;
  margin: 0;
  margin: 0 0.4rem 0.4rem 0;
}

<div>
  <span
    class="emotion-0"
  >
    hello
  </span>
</div>
`;

exports[`Tag should not output styles when unstyled is set 1`] = `
<div>
  <span
    class="emotion-0"
  >
    hello
  </span>
</div>
`;
